service/mapper/transitions.go

```diff
@@ -0,0 +1,334 @@
+// Copyright 2021 Optakt Labs OÜ
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may not
+// use this file except in compliance with the License. You may obtain a copy of
+// the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations under
+// the License.
+
+package mapper
+
+import (
+	"bytes"
+	"fmt"
+	"sort"
+	"sync"
+
+	"github.com/onflow/flow-go/ledger"
+	"github.com/onflow/flow-go/ledger/complete/mtrie/trie"
+	"github.com/onflow/flow-go/model/flow"
+
+	"github.com/optakt/flow-dps/models/index"
+)
+
+type TransitionFunc func(*State) error
+
+type Transitions struct {
+	cfg   Config
+	load  Loader
+	chain Chain
+	feed  Feeder
+	index index.Writer
+	once  *sync.Once
+}
+
+func NewTransitions(load Loader, chain Chain, feed Feeder, index index.Writer, options ...func(*Config)) *Transitions {
+
+	cfg := DefaultConfig
+	for _, option := range options {
+		option(&cfg)
+	}
+
+	t := Transitions{
+		cfg:   cfg,
+		load:  load,
+		chain: chain,
+		feed:  feed,
+		index: index,
+		once:  &sync.Once{},
+	}
+
+	return &t
+}
+
+func (t *Transitions) BootstrapState(s *State) error {
```

Comment:
I'm sure there's a reason to why you ordered these functions this way :D 

If my understanding of the flow is correct (my understanding is showcased in the image below), something like this makes more sense for me reading the file:

1. Bootstrap()
2. IndexChain()
3. IndexTree()
4. ForwardBlock()
5. UpdateTree() - this one is out of place but heck, what are you gonna do 😄 

Flow chart is a bit weird since there's a state that's not a state (`active`) and states that are not states (`matched` and `ready`).

![fsm (1)](https://user-images.githubusercontent.com/20494379/123985091-78004380-d9c5-11eb-88ce-cd8bde3744b3.jpg)

Reply:
This is outdated. However, to comment on the order, the natural flow is: update the tree until a finalized block commit is found, then index the payloads we collected, then forward the height of the block and repeat.

The bootstrapping is an edge case, which bypasses the first few steps for a number of reasons. However, the most accurate depiction of this would be a linear cyclical order like this:

```
updating => matched => indexed => forwarded =>
```

... with an `empty` being before the `updating`, but the pointer passing `matched` and `indexed` and going directly to `forwarded`.